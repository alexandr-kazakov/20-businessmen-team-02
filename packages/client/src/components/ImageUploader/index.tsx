import React, { useEffect, useRef, useState } from 'react'

import { Button, ButtonVariant } from '../../components/UI/Button'
import { showSnackBar } from '../..//components/Snackbar/redux/snackbarSlice'
import { useAppDispatch } from '../../app/redux/hooks'

import styles from './styles.module.scss'

export interface ImageUploaderProps {
  className?: string
  style?: React.CSSProperties
  onChange: (file: File | null) => void
  value?: File | null
  initPreview: string
  onClick: () => void
}

const IMAGE_MIME_TYPE = /image\/(png|jpg|jpeg)/i
const IMAGE_MAX_SIZE = 1000000

export const ImageUploader: React.FC<ImageUploaderProps> = ({
  className,
  style,
  onChange,
  value = null,
  initPreview,
  onClick,
}) => {
  const [imageFile, setImageFile] = useState<File | null>(value)
  const [imagePreview, setImagePreview] = useState('')
  const inputRef = useRef<HTMLInputElement>(null)
  const dispatch = useAppDispatch()

  const setSnackbarError = (message: string) => {
    dispatch(showSnackBar(message))
    throw new Error(message)
  }

  const onFileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { files } = e.target

    if (files && files.length) {
      const file = files[0]

      if (file.size > IMAGE_MAX_SIZE) {
        setSnackbarError('Максимальный размер файла 1MB')

        return
      }

      if (!file.type.match(IMAGE_MIME_TYPE)) {
        setSnackbarError(`Выбран не валидный формат файла: ${file.type}`)

        return
      }

      setImageFile(file)
      onChange(file)
    }
  }

  const onRemove = () => {
    setImageFile(null)
    setImagePreview('')
    onChange(null)
  }

  useEffect(() => {
    let isMounted = true
    let fileReader: FileReader

    if (imageFile) {
      // читаем содержимое файла для превью изображения
      new Promise<string>((resolve, reject) => {
        const reader = new FileReader()

        fileReader = reader

        reader.addEventListener('load', () => {
          if (reader.result && typeof reader.result === 'string') {
            resolve(reader.result)
          }
        })

        reader.addEventListener('abort', () => {
          reject(new Error('Чтение файла прервано'))
        })

        reader.addEventListener('error', () => {
          reject(new Error('Ошибка при чтении файла'))
        })

        reader.readAsDataURL(imageFile)
      })
        .then(image => isMounted && setImagePreview(image))
        .catch(e => console.error(e.message))
    }

    return () => {
      isMounted = false

      // если компонент отмонтирован а операция чтения была вызвана и не завершилась то прерываем ее
      if (fileReader && fileReader.readyState === 1) {
        fileReader.abort()
      }
    }
  }, [imageFile])

  const onSave = async () => {
    await onClick()
    setImagePreview('')
    setImageFile(null)
  }

  return (
    <div className={className}>
      {!imagePreview && (
        <label tabIndex={0} className={styles.upload} style={style}>
          <img
            className={styles.image}
            src={
              initPreview
                ? `https://ya-praktikum.tech/api/v2/resources/${initPreview}`
                : 'https://images.unsplash.com/photo-1633332755192-727a05c4013d?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1480&q=80'
            }
            alt=""
          />
          <input
            className={styles.input}
            onChange={onFileChange}
            ref={inputRef}
            accept="image/jpeg, image/jpg, image/png"
            type="file"
          />
        </label>
      )}

      {imagePreview && (
        <div className={styles.preview}>
          <img className={styles.image} src={imagePreview} alt="" />
          <Button variant={ButtonVariant.SECONDARY} className={styles.close} onClick={onRemove}>
            Отменить
          </Button>
          <Button variant={ButtonVariant.SECONDARY} className={styles.save} onClick={onSave}>
            Сохранить
          </Button>
        </div>
      )}
    </div>
  )
}
