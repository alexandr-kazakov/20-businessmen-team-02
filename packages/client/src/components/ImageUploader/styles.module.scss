.upload {
  background-color: transparent;
  cursor: pointer;
  display: flex;
  justify-content: center;

  & .input {
    pointer-events: none;
    position: absolute;
    display: none;
  }
}

.image {
  width: 200px;
  height: 200px;
  min-width: 200px;
  border-radius: 50%;
  object-fit: cover;
  overflow: hidden;
}

.close {
  cursor: pointer;
}

.save {
  cursor: pointer;
}

.preview {
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
}
