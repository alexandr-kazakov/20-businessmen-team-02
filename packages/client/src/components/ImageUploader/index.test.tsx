import React from 'react'
import { render } from '@testing-library/react'
import { ImageUploader } from '@/components/ImageUploader/index'

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}))

test('should render ImageUploader component', () => {
  const { asFragment } = render(<ImageUploader onChange={jest.fn()} onClick={jest.fn()} initPreview="example.img" />)

  expect(asFragment()).toMatchSnapshot()
})
