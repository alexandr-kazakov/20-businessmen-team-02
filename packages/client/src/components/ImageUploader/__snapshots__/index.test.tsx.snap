// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ImageUploader component 1`] = `
<DocumentFragment>
  <div>
    <label
      tabindex="0"
    >
      <img
        alt=""
        src="https://ya-praktikum.tech/api/v2/resources/example.img"
      />
      <input
        accept="image/jpeg, image/jpg, image/png"
        type="file"
      />
    </label>
  </div>
</DocumentFragment>
`;
